modules/lang-painless/build.gradle

```diff
@@ -49,17 +52,41 @@ task apiJavadoc(type: Javadoc) {
   include '**/org/elasticsearch/painless/api/'
   destinationDir = new File(docsDir, 'apiJavadoc')
 }
+
 task apiJavadocJar(type: Jar) {
   classifier = 'apiJavadoc'
   from apiJavadoc
 }
+
 assemble.dependsOn apiJavadocJar
 
-// Reference documentation for Painless's public API.
-task generatePainlessApi(type: JavaExec) {
-  main = 'org.elasticsearch.painless.PainlessDocGenerator'
-  classpath = sourceSets.test.runtimeClasspath
-  args file('../../docs/painless/painless-api-reference')
+/**********************************************
+ *           Context API Generation           *
+ **********************************************/
+
+sourceSets {
+  doc {
+    java {
+      srcDirs = ['src/doc/java']
+    }
+  }
+}
+
+dependencies {
+  docCompile "org.elasticsearch:elasticsearch:${project.versions.elasticsearch}"
+  docCompile project(':modules:lang-painless')
+}
+
+ClusterConfiguration clusterConfig = project.extensions.create("generateContextCluster", ClusterConfiguration.class, project)
+gradle.projectsEvaluated {
+  project.ext.generateContextNodes = ClusterFormationTasks.setup(project, "generateContextCluster", generateContextDoc, clusterConfig)
+}
+clusterConfig.distribution = 'default'
+
+task generateContextDoc(type: JavaExec) {
```

Comment:
If we could add proper input/outputs to this task so we could benefit from UP-TO-DATE checking and eventually the build cache that would be great.

https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_runtime_api

Reply:
I would prefer to defer this to a different PR for now.  This task is run separately from the standard build and is normally not be run in general. The infrastructure to effectively check the inputs (new script contexts, changes to resource file whitelists, and changes to the code generation) isn't fully supported, yet.  Once this task is integrated as part of the standard build, I would be happy to re-address this.